src/main/java/app/linguistai/bmvp/service/gamification/FriendshipService.java

```diff
@@ -25,84 +31,100 @@ public class FriendshipService {
     @Transactional
     public Friendship sendFriendRequest(String user1Email, UUID user2Id) throws Exception {
         try {
-            User dbUser1 = accountRepository.findUserByEmail(user1Email).orElseThrow(() -> new Exception("User is not found"));
-            User dbUser2 = accountRepository.findUserById(user2Id).orElseThrow(() -> new Exception("Requested user is not found"));
+            User dbUser1 = accountRepository.findUserByEmail(user1Email).orElseThrow(() -> new NotFoundException("User", true));
+            User dbUser2 = accountRepository.findUserById(user2Id).orElseThrow(() -> new NotFoundException("Requested user", true));
 
             if (dbUser1.getId().equals(dbUser2.getId())) {
-                throw new Exception("User cannot send friend request to themselves");
+                throw new FriendException("You cannot send friend request to yourself");
             }
 
             // check if friend request exists
             Friendship friendship = friendshipRepository.findByUserPair(dbUser1.getId(), user2Id).orElse(null);
 
             if (friendship != null) {
-                throw new Exception("Friendship or request already exist");
+                if (friendship.getStatus() == FriendshipStatus.PENDING) {
+                    throw new AlreadyFoundException("Friendship request", true);
```

Comment:
Here, the code checks `friendship.getStatus()` and then based on the result, throws different exceptions to differentiate between the cases. This introduces nesting and also a little bit of repetition.

Utilizing `switch expressions` introduced in Java 14 will significantly reduce nesting and improve the readability and maintainability of the code. Example:

```
switch (number) {
	case 1 -> callOne();
	case 2 -> callTwo();
	default -> callMany();
}
```

See [https://nipafx.dev/java-switch/](https://nipafx.dev/java-switch/) for details.